cmake/target_link_messages.cmake

```diff
@@ -0,0 +1,53 @@
+# Copyright 2023 Open Source Robotics Foundation, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+##################################################
+# A function that does whole-archive linking for messages
+function(target_link_messages)
+  set(options PUBLIC PRIVATE)
+  set(oneValueArgs TARGET)
+  set(multiValueArgs MSG_TARGETS)
+
+  cmake_parse_arguments(target_link_messages "${options}" "${oneValueArgs}" "${multiValueArgs}" ${ARGN})
+
+  if (target_link_messages_PUBLIC)
+    set(VISIBILITY PUBLIC)
+  elseif (target_link_messages_PRIVATE)
+    set(VISIBILITY PRIVATE)
+  endif()
+
+  foreach(message_lib ${target_link_messages_MSG_TARGETS})
+    if (CMAKE_CXX_COMPILER_ID STREQUAL "MSVC")
+      # MSVC link flag doesn't work with generator expressions
+      # TODO(mjcarroll) When CMake 3.24 is genrally available, use
+      # linking generator expressions as described here:
+      # https://cmake.org/cmake/help/latest/manual/cmake-generator-expressions.7.html#genex:LINK_LIBRARY
+      target_link_libraries(${target_link_messages_TARGET} ${VISIBILITY} -WHOLEARCHIVE:$<TARGET_FILE:${message_lib}>)
+
+      # Explicitly add dependency, link libraries, and includes as the WHOLEARCHIVE flag doesn't 
+      # do that on Windows.
+      add_dependencies(${target_link_messages_TARGET} ${message_lib})
+      get_target_property(message_lib_INCLUDES ${message_lib} INTERFACE_INCLUDE_DIRECTORIES)
+      get_target_property(message_lib_LIBS ${message_lib} INTERFACE_LINK_LIBRARIES)
+      target_include_directories(${target_link_messages_TARGET} PRIVATE ${message_lib_INCLUDES})
+      target_link_libraries(${target_link_messages_TARGET} ${message_lib_LIBS})
```

Comment:
I remember now that we can't use visibility here, because gz_add_tests uses the `target_link_libraries` function with out visibility.

If we try to link the messages into a test, we get cmake errors such as:

```
  CMake Error at cmake/target_link_messages.cmake:62 (target_link_libraries):
    The plain signature for target_link_libraries has already been used with
    the target "UNIT_PointCloudPackedUtils_TEST".  All uses of
    target_link_libraries with a target must be either all-keyword or
    all-plain.
  
    The uses of the plain signature are here:
  
     * /usr/share/cmake/gz-cmake3/cmake3/GzBuildExecutables.cmake:112 (target_link_libraries)
     * /usr/share/cmake/gz-cmake3/cmake3/GzBuildExecutables.cmake:116 (target_link_libraries)
     * /usr/share/cmake/gz-cmake3/cmake3/GzBuildTests.cmake:144 (target_link_libraries)     ```